lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c

```diff
@@ -2059,6 +2059,326 @@ int s1ap_mme_handle_handover_request_ack(
+    const sctp_stream_id_t stream, S1ap_S1AP_PDU_t* pdu) {
+  S1ap_HandoverFailure_t* container = NULL;
+  S1ap_HandoverFailureIEs_t* ie     = NULL;
+  S1ap_S1AP_PDU_t out_pdu           = {0};
+  S1ap_HandoverPreparationFailure_t* out;
+  S1ap_HandoverPreparationFailureIEs_t* hpf_ie = NULL;
+  ue_description_t* ue_ref_p                   = NULL;
+  mme_ue_s1ap_id_t mme_ue_s1ap_id              = INVALID_MME_UE_S1AP_ID;
+  S1ap_Cause_PR cause_type;
+  long cause_value;
+  uint8_t* buffer_p = NULL;
+  uint8_t err       = 0;
+  uint32_t length   = 0;
+
+  OAILOG_FUNC_IN(LOG_S1AP);
+
+  container = &pdu->choice.unsuccessfulOutcome.value.choice.HandoverFailure;
+
+  // get the mandantory IEs
+  // MME_UE_S1AP_ID
+  S1AP_FIND_PROTOCOLIE_BY_ID(
+      S1ap_HandoverFailureIEs_t, ie, container,
+      S1ap_ProtocolIE_ID_id_MME_UE_S1AP_ID, true);
+  if (ie) {
+    mme_ue_s1ap_id = ie->value.choice.MME_UE_S1AP_ID;
+  } else {
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+
+  // Grab the Cause Type and Cause Value
+  S1AP_FIND_PROTOCOLIE_BY_ID(
+      S1ap_HandoverFailureIEs_t, ie, container, S1ap_ProtocolIE_ID_id_Cause,
+      true);
+  if (ie) {
+    cause_type = ie->value.choice.Cause.present;
+  } else {
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+  switch (cause_type) {
+    case S1ap_Cause_PR_radioNetwork:
+      cause_value = ie->value.choice.Cause.choice.radioNetwork;
+      break;
+
+    case S1ap_Cause_PR_transport:
+      cause_value = ie->value.choice.Cause.choice.transport;
+      break;
+
+    case S1ap_Cause_PR_nas:
+      cause_value = ie->value.choice.Cause.choice.nas;
+      break;
+
+    case S1ap_Cause_PR_protocol:
+      cause_value = ie->value.choice.Cause.choice.protocol;
+      break;
+
+    case S1ap_Cause_PR_misc:
+      cause_value = ie->value.choice.Cause.choice.misc;
+      break;
+
+    default:
+      OAILOG_ERROR(
+          LOG_S1AP, "HANDOVER FAILURE with Invalid Cause_Type = %d\n",
+          cause_type);
+      OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+
+  // A failure means the target eNB was unable to prepare for handover. We need
+  // to get rid of UE handover state and send a failure message with cause back
+  // to the source eNB.
+
+  // get UE context
+  if ((ue_ref_p = s1ap_state_get_ue_mmeid(mme_ue_s1ap_id)) == NULL) {
+    OAILOG_ERROR(
+        LOG_S1AP,
+        "could not get ue context for mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT
+        ", failing!\n",
+        (uint32_t) mme_ue_s1ap_id);
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+
+  if (ue_ref_p->s1_ue_state == S1AP_UE_HANDOVER) {
+    // this effectively cancels the HandoverPreparation proecedure as we
+    // only send a HandoverCommand if the UE is in the S1AP_UE_HANDOVER
+    // state.
+    ue_ref_p->s1_ue_state         = S1AP_UE_CONNECTED;
+    ue_ref_p->s1ap_handover_state = (struct s1ap_handover_state_s){0};
+  } else {
+    // Not a failure, but nothing for us to do.
```

Comment:
do we need to continue with the processing of the message? What is the side effect or how do we end up in this state?

Reply:
This is about cleaning up the state on the source eNB. We can get into this situation if:

1. Source eNB sends a HandoverRequired to the MME
2. The MME sends the HandoverRequest to the Target eNB
3. The Source eNB sends a HandoverCancel (for some reason) to the MME, which is processed.
4. The Target eNB (for whatever reason) separately rejects the handover, generating a failure.

It'd be reasonable to assume that most eNBs, after sending a cancel, will clean up their own state and ignore the HandoverPreparationFailure message we send, but my read of the spec suggests that we should still send the unsuccessful outcome message to properly complete the procedure just in case.

I am sure there are other situations where the UE may change state (e.g., power off) during the handover procedure as well. In all cases, we should still notify the source eNB that the target eNB did not accept the handover.

Reply:
sounds good.